appsmithorg/appsmith
# fix: changed validation to FUNCTION for the default selected value RadioGroupWidget

## Description

**Summary of changes**

- Changed the validation to function type for `defaultOptionValue`.
- Added custom validator function for options property
    - This also checks that all the data types of the `value` property inside options is similar. 
- Changed the logic for onRadioSelectionChange function. 
    - Checking the data type of `value` property of the first object inside options property to be number. If true, parsing it to float.
    - Else keeping value as is.

**Motivation**

- Evaluated value for the defaultOptionValue is always the string irrespective of the mustasch syntax i.e. {{}}. This caused issue when the value is integer in the `value` property of the `options` property. Now with changes included in this PR -> value inside mustach is treated as integer if passed integer e.g. {{1}} <-> int. And normal value is treated as string e.g. 1 is string. 
-  Whenever the `value` property inside the `options` property has integer values then, clicking on radio button does not change the state to checked. This issue also got fixed in this PR.

Fixes #9834 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Added the following tests:

- Cypress test: `RadioGroup_Int_Value_spec.js`,
    - This spec file checks the following
        - If the data type of the value in the `value` property is integer or string, do the following checks:
        - Check if the default selected value is changing the radio but as expected.
        - Check if the radio buttons can be clicked normally.
        - Checks if all the data types of the value property in the options property is the same.
- Unit test: `propertyPaneConfig.test.ts`
    - This test checks the defaultOptionValue custom validator. It checks if number passed to the validator returns parsed value as an integer. Same the case with string. 

## Checklist:

















## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>


    // Code coverage diff between base branch:release and head branch: fix/fix-radio_group_value_to_integer_map 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.06 **(-0.05)** | 36.84 **(-0.03)** | 34.99 **(0)** | 55.55 **(-0.05)**
 :red_circle: | app/client/src/selectors/commentsSelectors.ts | 83.61 **(-1.64)** | 61.76 **(-2.95)** | 73.33 **(0)** | 88.24 **(-2.35)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**
 :red_circle: | app/client/src/widgets/RadioGroupWidget/widget/index.tsx | 33.77 **(-50.85)** | 6.06 **(6.06)** | 63.64 **(-11.36)** | 34.72 **(-52.78)**</details>
